File: CS3305/Lab03/C++/Lab03C++/sequence_test.cxx
// FILE: sequence_test.cxx
// An interactive test program for the new sequence class
#include <cctype>       // Provides toupper
#include <iostream>     // Provides cout and cin
#include <cstdlib>      // Provides EXIT_SUCCESS and system
#include "sequence.h"   // With value_type defined as double
using namespace std;

typedef sequence::size_type size_type;
typedef sequence::value_type value_type;

// PROTOTYPES for functions used by this test program:
void print_menu( );
// Postcondition: A menu of choices for this program has been written to cout.

char get_user_command( );
// Postcondition: The user has been prompted to enter a one character command.
// The next character has been read (skipping blanks and newline characters), 
// and this character has been returned.

void show_sequence(const sequence& display);
// Postcondition: The items on display have been printed to cout.

double get_number( );
// Postcondition: The user has been prompted to enter a number. The
// number has been read, echoed to the screen, and returned by the function.

size_t get_whole_number( );
// Postcondition:  The user has been prompted to enter a non-negative whole number.
// The number has been read, echoed to the screen, and returned by the function.

void pressEnterToContinue( );



int main( ) {
    sequence test1, test2, test3; // Sequences that we'll perform tests on
    char choice;   // A command character entered by the user
    
    system("clear");
    cout << "I have initialized empty sequences of numbers." << endl;
    cout << "\tPress enter to continue . . . ";
    cin.ignore();

    do {
        system("clear");
        print_menu( );
        choice = get_user_command( );
        switch (choice) {
            case 's': test1.start( );
                      break;
            case 'a': if (test1.is_current())
                          test1.advance( );
                      else
                          cout << "Cannot advance since cursor is not current" << endl;
                      pressEnterToContinue();
                      break;
            case '?': if (test1.is_current( ))
                          cout << "There is an item." << endl;
                      else 
                          cout << "There is no current item." << endl;
                      pressEnterToContinue();
                      break;
            case 'c': if (test1.is_current( ))
                           cout << "Current item is: " << test1.get_current( ) << endl;
                      else
                          cout << "There is no current item." << endl;
                      pressEnterToContinue();
                      break;
            case '1': show_sequence(test1);
                      pressEnterToContinue();
                      break;
            case '2': show_sequence(test2);
                      pressEnterToContinue();
                      break;
            case '3': show_sequence(test3);
                      pressEnterToContinue();
                      break;
            case 'S': cout << "Size is " << test1.size( ) << '.' << endl;
                      pressEnterToContinue();
                      break;
            case 'I': test1.add_before(get_number( ));
                      break;
            case 'A': test1.add_after(get_number( ));
                      break;
            case 'R': test1.remove_current( );
                      cout << "The current item has been removed." << endl;
                      pressEnterToContinue();
                      break;   
            case '=': test2 = test1;
                      break;
            case '+': test3 = test1 + test2;
                      break;
            case 'e': cout << "test1 == test3? " << (test1==test3 ? "true" : "false") << endl;
                      pressEnterToContinue();
                      break;
            case 'i': test1.ensure_capacity(get_whole_number());
                      break;
            case 't': test1.trim_to_size();
                      break;
            case 'C': test3 = sequence(test1);
                      break;
            case 'q':
            case 'Q': cout << "\t\tRidicule is the best test of truth." << endl;
                      pressEnterToContinue();
                      break;
            default:  cout << "\t\t" << choice << " is invalid." << endl;
                      pressEnterToContinue();
                      break;
        }
    }
    while ((choice != 'Q') && (choice != 'q'));

    return EXIT_SUCCESS;
}

void print_menu( ) {
// Library facilities used: iostream
    cout << endl; // Print blank line before the menu
    cout << "\n\t\tThe following choices are available: " << endl;
    cout << "\t\t s   Activate the start( ) function for test1" << endl;
    cout << "\t\t a   Activate the advance( ) function for test1" << endl;
    cout << "\t\t ?   Print the result from the is_item( ) for test1" << endl;
    cout << "\t\t c   Print the result from the current( ) for test1" << endl;
    cout << "\t\t 1   Print a copy of test1" << endl;
    cout << "\t\t 2   Print a copy of test2" << endl;
    cout << "\t\t 3   Print a copy of test3" << endl;
    cout << "\t\t S   Print the result from the size( ) function for test1" << endl;
    cout << "\t\t I   Insert a new number with the insert(...) function for test1" << endl;
    cout << "\t\t A   Attach a new number with the attach(...) function for test1" << endl;
    cout << "\t\t R   Activate the remove_current( ) function for test1" << endl;
    cout << "\t\t =   Assign test1 to test2" << endl;
    cout << "\t\t +   Assign test3 expression test1 + test2" << endl;
    cout << "\t\t e   Check whether test1 == test3" << endl;
    cout << "\t\t i   Ensure capacity on test1" << endl;
    cout << "\t\t t   Trim to size on test1" << endl;
    cout << "\t\t C   Use copy constructor test3 = sequence(test1)" << endl;
    cout << "\t\t Q   Quit this test program" << endl;
}

char get_user_command( ) {
// Library facilities used: iostream
    char command;

    cout << "\t\tEnter choice: ";
    cin >> command; // Input of characters skips blanks and newline character

    return command;
}

void show_sequence(const sequence& display) {
// Library facilities used: iostream
    cout << display << endl;
}

double get_number( ) {
// Library facilities used: iostream
    value_type result;
    
    cout << "Please enter a number for the sequence: ";
    cin  >> result;
    cout << result << " has been read." << endl;
    return result;
}

size_t get_whole_number( ) {
// Library facilities used:  iostream
    size_t result;

    cout << "\t\tPlease enter a non-negative whole number: ";
    cin >> result;
    cout << "\t\t" << result << " has been read." << endl;
    return result;
}

void pressEnterToContinue( ) {
// Library facilities used:  iostream
    char ch;
    cout << "\n\t\tPress enter to continue  . . . ";
    cin.ignore(); cin.ignore();
}

File: CS3305/Lab05/Java/Handout/SequenceTest.java
import java.util.Scanner;

public class SequenceTest {

    public static Scanner si = new Scanner( System.in ); //tie scanner to stdin

    public static void main( String [] args ) {

        DLinkedSequence<Long> test1, test2, test3;
        test1 = new DLinkedSequence<Long>();
        test2 = new DLinkedSequence<Long>();
        test3 = new DLinkedSequence<Long>();

        char choice = ' ';
        long number = 0L;

        System.out.printf("\033c"); //clear screen
        System.out.printf("I have initialized three empty sequences of numbers: test1, test2, and test3.\n");
        pressEnterToContinue();

        do {
            System.out.printf("\033c"); //clear screen
            print_menu();
            choice = get_user_command();

            switch (choice) {
                case 's' : test1.begin();
                           break;
                case 'a' : if ( test1.isCurrent() ) 
                               test1.advance();
                           else {
                               System.out.printf("Cannot advance since cursor is not current\n");
                               pressEnterToContinue();
                           }
                           break;
                case 'r' : if ( test1.isCurrent() ) 
                               test1.retreat();
                           else {
                               System.out.printf("Cannot retreat since cursor is not current\n");
                               pressEnterToContinue();
                           }
                           break;
                case '?' : if ( test1.isCurrent() )
                               System.out.printf("There is a current item\n");
                           else 
                               System.out.printf("There is no current item\n");
                           pressEnterToContinue();
                           break;
                case 'g' : if ( test1.isCurrent() )
                               System.out.printf("Current item is: %d\n", test1.getCurrent());
                           else
                               System.out.printf("There is no current element\n");
                           pressEnterToContinue();
                           break;
                case '1' : showSequence( test1 );
                           pressEnterToContinue();
                           break;
                case '2' : showSequence( test2 );
                           pressEnterToContinue();
                           break;
                case '3' : showSequence( test3 );
                           pressEnterToContinue();
                           break;
                case 'S' : System.out.printf("Size of test1 is: %d\n", test1.size());
                           pressEnterToContinue();
                           break;
                case 'I' : test1.addBefore( get_long_number() );
                           break;
                case 'A' : test1.addAfter( get_long_number() );
                           break;
                case 'R' : if ( test1.isCurrent() ) {
                               test1.removeCurrent();
                               System.out.printf("test1's current item has been removed\n");
                           } 
                           else System.out.printf("test1's cursor not current -- not removed\n");
                           pressEnterToContinue();
                           break;
                case '+' : showSequence( test1.concatenation( test3 ) );
                           pressEnterToContinue();
                           break;
                case 'e' : System.out.printf("test1.equals(test3)? %s\n", test1.equals(test3) );
                           pressEnterToContinue();
                           break;
                case 'C' : test3 = new DLinkedSequence<Long>( test1 );
                           break;
                case 'c' : test3 = test1.clone();
                           break;
                case 'q' :
                case 'Q' : System.out.printf("\t\tRidicule is the best test of truth\n");
                           pressEnterToContinue();
                           break;
                default  : System.out.printf("\t\t%c is invalid . . . \n", choice);
                           pressEnterToContinue();
                           break;
            }
        } while ( (choice != 'Q') && (choice != 'q') );

    }

    public static void print_menu( ) {
        System.out.printf("\n\n\t\tThe following choices are available: \n");
        System.out.printf("\t\t s  Activate the start() method for test1\n");
        System.out.printf("\t\t a  Activate the advance() method for test1\n");
        System.out.printf("\t\t r  Activate the retreat() method for test1\n");
        System.out.printf("\t\t ?  Display the result from the isCurrent() for test1\n");
        System.out.printf("\t\t g  Display the result from the getCurrent() for test1\n");
        System.out.printf("\t\t 1  Display test1\n");
        System.out.printf("\t\t 2  Display test2\n");
        System.out.printf("\t\t 3  Display test3\n");
        System.out.printf("\t\t S  Display the result for the size() method for test1\n");
        System.out.printf("\t\t I  Add a new number to test1 with addBefore()\n");
        System.out.printf("\t\t A  Add a new number to test1 with addAfter()\n");
        System.out.printf("\t\t R  Remove current value from test1 with removeCurrent()\n");
        System.out.printf("\t\t +  Display the concatenation of test1 and test3\n");
        System.out.printf("\t\t e  Check whether test1.equals(test3)\n");
        System.out.printf("\t\t C  Use copy constructor test3 = new DLinkedSequence<Long>(test1)\n");
        System.out.printf("\t\t c  Use clone of test1 by test3 = test1.clone()\n");
        System.out.printf("\t\t Q  Quit this test program\n");
    }

    public static void pressEnterToContinue() {
        System.out.printf("\t\tPress Enter to continue . . .");
        try {
            System.in.read();
        } catch(Exception e) { }
    }

    public static void showSequence( DLinkedSequence<Long> seq ) {
        System.out.printf("%s\n", seq);
    }

    public static char get_user_command( ) {
        char command = ' ';

        System.out.printf("\tEnter choice: ");
        try {
            if ( si.hasNext() ) 
                command = si.next( ).charAt(0);
            else
                command = (char) System.in.read();
        }
        catch ( Exception e ) {}
        return command;
    }

    public static double get_double_number( ) {
        double result;

        System.out.printf("\tPlease enter a double number: ");
        result = si.nextDouble();
        return result;
    }

    public static long get_long_number( ) {
        long result;
        System.out.printf("\tPlease enter a long number: ");
        result = si.nextLong();
        return result;
    }

    public static int get_int_number( ) {
        int result;
        System.out.printf("\tPlease enter an integer number: ");
        result = si.nextInt();
        return result;
    }


    public static char toupper( char aChar ) {
        if ( aChar >= 'a' && aChar <= 'z' ) {
            int temp = (int)aChar ^ 0x20;
            aChar = (char)temp;
        }
        return aChar;
    }

}

